dataset/slice.cpp

```diff
@@ -0,0 +1,101 @@
+                         const VariableConstView begin,
+                         const VariableConstView end) {
+  using namespace variable;
+  if (begin)
+    variable::detail::expect0D(begin.dims());
+  if (end)
+    variable::detail::expect0D(end.dims());
+  auto coords = to_slice.coords();
+  if (!coords.contains(dim))
+    throw except::DimensionNotFoundError(to_slice.dims(), dim);
+  auto [k, coord] = *coords.find(dim);
+  if (coord.dims().ndim() != 1) {
+    throw except::SizeError(
+        "Multi-dimensional coordinates not supported in slice");
+  }
+  const bool ascending = is_sorted(coord, dim, variable::SortOrder::Ascending);
+  const bool descending =
+      is_sorted(coord, dim, variable::SortOrder::Descending);
+
+  if (!(ascending ^ descending))
+    throw std::runtime_error("Coordinate must be monotomically increasing or "
+                             "decreasing for value slicing");
+
+  const auto len_data = to_slice.dims()[dim];
+  const auto len = coord.dims().volume();
+  const auto bin_edges = len == len_data + 1;
+
+  // Point slice
+  if ((begin && end) && begin == end) {
+    scipp::index idx = -1;
+    if (bin_edges) {
+      if (ascending)
+        idx = sum(less_equal(coord, begin), dim).value<scipp::index>() - 1;
+      else
+        idx = sum(greater_equal(coord, begin), dim).value<scipp::index>() - 1;
+      if (idx < 0 || idx >= len_data)
+        throw except::NotFoundError(
+            to_string(begin) +
+            " point slice does not fall within any bin edges along " +
+            to_string(dim));
+    } else {
+      auto eq = equal(coord, begin);
+      auto values_view = eq.values<bool>();
+      auto it = std::find(values_view.begin(), values_view.end(), true);
+      if (it == values_view.end())
+        throw except::NotFoundError(to_string(begin) +
+                                    " point slice does not exactly match any "
+                                    "point coordinate value along " +
+                                    to_string(dim));
+      idx = std::distance(values_view.begin(), it);
+    }
+    return to_slice.slice({dim, idx});
+  }
+  // Range slice
+  else {
+    scipp::index first = 0;
+    scipp::index last = len - 1;
+    if (begin) {
+      if (bin_edges) {
+        // lower bin edge boundary
+        if (ascending) {
+          first = sum(less_equal(coord, begin), dim).value<scipp::index>() - 1;
+        } else {
+          first =
+              sum(greater_equal(coord, begin), dim).value<scipp::index>() - 1;
+        }
+      } else {
+        // First point >= value for non bin edges
+        if (ascending)
+          first =
+              len - sum(greater_equal(coord, begin), dim).value<scipp::index>();
+        else
+          first =
+              len - sum(less_equal(coord, begin), dim).value<scipp::index>();
+      }
+      if (first < 0)
+        first = 0;
+    }
+    if (end) {
+      if (ascending) {
```

Comment:
Massive duplication around here. There should not be so many cases? For example, finding the index of bins is the same, regardless of whether it for begin or end (just need to add 1 to the index of the last bin). Also finding indices for non-bin case is the same, just the intepretation of the indices changes?